torch/_dynamo/guards.py

```diff
@@ -526,6 +541,153 @@ def guard_on_dict_keys_and_order(self, value, guard):
                     key, get_verbose_code_parts(f"{key_source} == {key!r}", guard)
                 )
 
+    def getattr_on_nn_module(
+        self,
+        source,
+        base_guard_manager,
+        base_example_value,
+        example_value,
+        base_source_name,
+        source_name,
+        guard_manager_enum,
+    ):
+        """
+        This tries to avoid calling the expensive nn module custom getattr method by
+        checking if the attribute is accessible via __dict__. For attributes that
+        are not accessible via __dict__ (like descriptors), we fallback to
+        PyObject_GetAttr.
+
+        There are two cases that we optimize for
+        1) attributes present directly in __dict__, e.g training.
+        2) parameters/buffers/modules - they can be accessed via _parameters,
+        _buffers, _modules keys in __dict__. For example, mod.linear can be
+        accessed as mod.__dict__["_parameters"]["linear"]
+
+        The most common and expensive case for nn module guards is of type
+        mod.submod1.submod2.submod3.training. We avoid the python getattr of nn
+        modules by going through the __dict__.
+        """
+
+        def getitem_on_dict_mgr(
+            mgr, key, source_name, base_example_value, example_value, guard_manager_enum
+        ):
+            if isinstance(mgr, DictGuardManager):
+                # Case where the user code relies on key order, e.g.,
+                # named_parameters
+                index = get_key_index(base_example_value, key)
+
+                # Install the key manager and add equals match guard
+                key_source = f"list({source_name}.keys())[{index!r}]"
+                mgr.get_key_manager(
+                    index=index,
+                    source=key_source,
+                    example_value=key,
+                    guard_manager_enum=GuardManagerType.GUARD_MANAGER,
+                ).add_equals_match_guard(l2_key, [f"{key_source} == {l2_key!r}"])
+
+                # Install the value manager
+                return mgr.get_value_manager(
+                    index=index,
+                    source=source_name,
+                    example_value=example_value,
+                    guard_manager_enum=guard_manager_enum,
+                )
+            else:
+                return mgr.dict_getitem_manager(
+                    key=key,
+                    source=source_name,
+                    example_value=example_value,
+                    guard_manager_enum=guard_manager_enum,
+                )
+
+        attr_name = source.member
+        mod_dict = base_example_value.__dict__
+
+        all_class_attribute_names: Set[str] = set()
+        for x in inspect.getmro(base_example_value.__class__):
+            all_class_attribute_names.update(x.__dict__.keys())
+
+        if attr_name in all_class_attribute_names:
```

Comment:
class attributes are shadowed by module dict attributes... I think the order of these are wrong.  add some tests for shadowing.

Reply:
You are right .. but there is a corner case with submodules, where class attr take precedence. This is the example

~~~
import torch
import torch._dynamo.testing

torch._dynamo.config.guard_nn_modules = True


class Mod(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.a = 3

    def forward(self,x, c=4):
        return x * c

    def linear(self, x):
        return x

    # b is class attribute and this is a bug that jansel pointed out
    # mod.b overshadows Mod.b
    def b(self, x):
        assert False
        return x

class MyMod(Mod):
    def __init__(self):
        super().__init__()
        # IF THIS LINE ACTUALLY CHAGES LINEAR, IT WOULD FAIL WITH SHAPE MISMATCH ERROR
        self.linear = torch.nn.Linear(11, 11)
        self.a = 2
        self.b = 2


    def forward(self, x, c=None):
        return self.linear(x) * self.a * self.b


mod = MyMod()
mod(torch.ones(3, 3))

# Instance attributes overwrite the instance attributes of the parent class.
assert mod.a == 2

# ANOMALY - mod.linear is not the parameter
assert mod.linear.__code__ is Mod.linear.__code__

# jansel's valid comment - module dict attr overshadow the class attrs
# anijain2305 - This does not cause a bug because today we fallback to getattr,
# but we are losing perf opportunity here.
assert mod.b == 2

cnts = torch._dynamo.testing.CompileCounter()
opt_mod = torch.compile(mod, backend=cnts)
opt_mod(torch.ones(3, 3))
opt_mod(torch.ones(3, 3))

assert cnts.frame_count == 1

~~~


The discrepancy is for `mod.linear`. Parent class has a function named `linear`. When the derived class instance calls self.linear, it calls the parent class function.

This is a special case. Apart from that everything else conforms to the general rule - "class attributes are shadowed by the mod dict attributes".

I also think this is a bug in `__setattr__` of nn.Module. For example, I am unable to assign a `self.param` if the parent class already has a function named `param`. If you think, this is just a bug, I can send a PR for nn modules.